share/spack/qa/setup-env-test.fish

```diff
@@ -0,0 +1,392 @@
+    echo_msg "---------------------------------"
+
+end
+
+# echo FAIL in red text; increment failures
+function fail
+    echo_red FAIL
+    set __spt_errors (math $__spt_errors+1)
+end
+
+# echo SUCCESS in green; increment successes
+function pass
+    echo_green SUCCESS
+    set __spt_success (math $__spt_success+1)
+end
+
+
+#
+# Run a command and suppress output unless it fails.
+# On failure, echo the exit code and output.
+#
+function spt_succeeds
+    printf "'$argv' succeeds ... "
+
+    set -l output (eval $argv 2>&1)
+
+    if test $status -ne 0
+        fail
+        echo_red "Command failed with error $status"
+        if test -n "$output"
+            echo_msg "Output:"
+            echo "$output"
+        else
+            echo_msg "No output."
+        end
+    else
+        pass
+    end
+end
+
+
+#
+# Run a command and suppress output unless it succeeds.
+# If the command succeeds, echo the output.
+#
+function spt_fails
+    printf "'$argv' fails ... "
+
+    set -l output (eval $argv 2>&1)
+
+    if test $status -eq 0
+        fail
+        echo_red "Command failed with error $status"
+        if test -n "$output"
+            echo_msg "Output:"
+            echo "$output"
+        else
+            echo_msg "No output."
+        end
+    else
+        pass
+    end
+end
+
+
+#
+# Ensure that a string is in the output of a command.
+# Suppresses output on success.
+# On failure, echo the exit code and output.
+#
+function spt_contains
+    set -l target_string $argv[1]
+    set -l remaining_args $argv[2..-1]
+
+    printf "'$remaining_args' output contains '$target_string' ... "
+
+    set -l output (eval $remaining_args 2>&1)
+
+    if not echo "$output" | string match -q -r ".*$target_string.*"
+        fail
+        echo_red "Command exited with error $status"
+        echo_red "'$target_string' was not in output."
+        if test -n "$output"
+            echo_msg "Output:"
+            echo "$output"
+        else
+            echo_msg "No output."
+        end
+    else
+        pass
+    end
+end
+
+
+#
+# Ensure that a variable is set.
+#
+function is_set
+    printf "'$argv[1]' is set ... "
+
+    if test -z "$$argv[1]"
+        fail
+        echo_msg "'$argv[1]' was not set!"
+    else
+        pass
+    end
+end
+
+
+#
+# Ensure that a variable is not set.
+# Fails and prints the value of the variable if it is set.
+#
+function is_not_set
+    printf "'$argv[1]' is not set ... "
+
+    if test -n "$$argv[1]"
+        fail
+        echo_msg "'$argv[1]' was set!"
+        echo "    $$argv[1]"
+    else
+        pass
+    end
+end
+
+
+
+# -----------------------------------------------------------------------
+# Setup test environment and do some preliminary checks
+# -----------------------------------------------------------------------
+
+# Make sure no environment is active
+set -e SPACK_ENV
+true # ingnore failing `set -e`
+
+# Source setup-env.sh before tests
+source share/spack/setup-env.fish
```

Comment:
This doesn't work unless the script is run from `SPACK_ROOT`. In #14393 I changed the test file to calculate which directory it lives in and find the right `setup-env` file.

Reply:
Yes, you're right. In the interest in getting this PR merged, I would vote to do this change in the follow-up PR (see previous comment above), when I add fish tab completion.

Reply:
Also adding this to my fork's issue list